File: day03/main.py
# https://adventofcode.com/2024/day/3
# https://xkcd.com/208/

import os
import re

def part1(lines: str, expected: int | None = None) -> int:
    # Start of implementation #
    pattern = re.compile(r"mul\((\d+),(\d+)\)")
    total = sum([int(m.group(1)) * int(m.group(2)) for m in pattern.finditer(lines)])
    # End of implementation #
    if expected is not None:
        assert total == expected
    return total


def part2(lines: str, expected: int | None = None) -> int:
    # Start of implementation #
    domul = re.compile(r"(do\(\)|don't\(\)|mul\(\d+,\d+\))")
    mul = re.compile(r"mul\((\d+),(\d+)\)")
    enabled = True
    total = 0
    for action in domul.finditer(lines):
        if action.group(0) == "do()":
            enabled = True
            continue
        elif action.group(0) == "don't()":
            enabled = False
            continue
        elif enabled:
            for m in mul.finditer(action.group(0)):
                total += int(m.group(1)) * int(m.group(2))

    # End of implementation #
    if expected is not None:
        assert total == expected
    return total


def process_file(filename: str, expected_part1: int | None = None, expected_part2: int | None = None):
    path = os.path.join(os.path.dirname(__file__), filename)
    with open(path, "r") as input_file:
        lines = input_file.read()

    p1 = part1(lines, expected_part1)
    p2 = part2(lines, expected_part2)
    return p1, p2


if __name__ == "__main__":
    example_part1_expected = 161
    example_part2_expected = 48
    ex1, ex2 = process_file("example", example_part1_expected, example_part2_expected)
    print("Example:")
    print(f"Part 1 expected {example_part1_expected} got {ex1}: {example_part1_expected == ex1}")
    print(f"Part 2 expected {example_part2_expected} got {ex2}: {example_part2_expected == ex2}")

    p1, p2 = process_file("input")
    print("Input:")
    print(f"Part 1: {p1}")
    print(f"Part 2: {p2}")

File: day03/src/main.rs
// https://adventofcode.com/2024/day/3
// https://xkcd.com/208/

use std::fs;
use regex::Regex;


fn part1(lines: &str, expected: Option<i32>) -> i32 {
    // Implementation here
    let re = Regex::new(r"mul\((\d+),(\d+)\)").unwrap();
    let mut total = 0;
    for (_, [first, second]) in re.captures_iter(lines).map(|c| c.extract()) {
        let first = first.parse().unwrap_or(0);
        let second = second.parse().unwrap_or(0);
        total += first * second;
    }
    if let Some(exp) = expected {
        println!("Part 1: Expected: {exp} Calculated: {total} Equal: {}", exp == total);
    }
    total
}

fn part2(lines: &str, expected: Option<i32>) -> i32  {
    // Implementation here
    let domulre = Regex::new(r"(do\(\)|don't\(\)|mul\(\d+,\d+\))").unwrap();
    let mulre = Regex::new(r"mul\((\d+),(\d+)\)").unwrap();
    let mut total = 0;
    let mut enabled = true;
    for (_, [action]) in domulre.captures_iter(lines).map(|c| c.extract()) {
        if action == "do()" {
            enabled = true;
            continue;
        }
        else if action == "don't()" {
            enabled = false;
            continue;
        }
        else if enabled {
            for (_, [first, second]) in mulre.captures_iter(action).map(|c| c.extract()) {
                let first = first.parse().unwrap_or(0);
                let second = second.parse().unwrap_or(0);
                total += first * second;
            }
        }
    }
    if let Some(exp) = expected {
        println!("Part 2: Expected: {exp} Calculated: {total} Equal: {}", exp == total);
    }
    total
}

fn process_file(path: &str, exp1: Option<i32>, exp2: Option<i32>) -> (i32, i32) {
    let input = fs::read_to_string(path).unwrap();
    //let lines: Vec<&str> = input.split('\n').into_iter().map(|l| {l.trim()}).collect();
    let part1_result = part1(&input, exp1);
    let part2_result = part2(&input, exp2);
    (part1_result, part2_result)
}

fn main() {
    let example_part1_expected = 161;
    let example_part2_expected = 48;
    let (ex1, ex2) = process_file("example", Some(example_part1_expected), Some(example_part2_expected));
    println!("Example 1: {ex1} 2: {ex2}");
    let (p1, p2) = process_file("input", None, None);
    println!("Input: part 1: {p1} part 2: {p2}");
}